### DISCUSSION TITLE
"Unable to compile template: 4uqjvolp.0oz(5,186): error CS0103: The name 'DateTime' does not exist in the current context"

### DISCUSSION THREAD

**Discussion by americanslon** [adoconnection/RazorEngineCore#119] [Q&A] [answered]
Here is the template

```
<br />
<b>Templating engine working correctly</b>
<br />
<div>
	I am loaded from cache with dynamically passed time. @ViewBag.Now
	<br />
	DateTime (proof assembly import working): @DateTime.Now
</div>
```

Here is the compilation code

```
IRazorEngineCompiledTemplate<ViewBagRazorEngineTemplate> template = engine.Compile<ViewBagRazorEngineTemplate>(templateText, builder =>
  {
      builder.AddAssemblyReferenceByName("System");
    
  });
```

When the engine tries to compile I get `The name 'DateTime' does not exist in the current context"` Removing `@DateTime.Now` from the template allows for the template to compile successfully. 

What am I missing?

Thank you!

**Comment by americanslon** [comment, marked as answer]
Pull my hair for two hours, post here and literally 5 minutes later get an idea for something to try and it turns out to be the problem,

For the next weary traveler

Code from the imported assemblies has to be fully qualified 

`DateTime (proof assembly import working): @DateTime.Now`
to
`DateTime (proof assembly import working): @System.DateTime.Now`

**Reply by adoconnection** [reply]
calling ```@DateTime``` is routed to public propery of template, similar to ```@Raw``` and ```@Model``` you may find it useful that you can introduce new keywords by extending ```RazorEngineTemplateBase``` and making public methods and properties. I use it all the time to make nice looking shortcuts like ```@Resource("abc")``` ```@Link(a, b, c)``` for your case it would be ```@Now```

see https://github.com/adoconnection/RazorEngineCore/blob/ce41fffe192b72c24be589c2c1ba007c1b2f2c32/RazorEngineCore.Tests/TestCompileAndRun.cs#L477

**Reply by americanslon** [reply]
Interesting.
In the similar vein, is there a way to specify `IRazorEngineCompiledTemplate` at run time? 
I wrapped RazorEngineCore in my own little extension, which's main purpose is to compile all the templates in a specified directory at the start of the application.

If there was a way to specify `IRazorEngineCompiledTemplate` at run time I could have the template type specified in the template itself and parse it out right before I call `.Compile`

That way I could have all these different templates inheriting from `RazorEngineTemplateBase` with all these different helpers in them.

**Reply by adoconnection** [reply]
Hi, sorry for late reply 😅 can you provide more info on what you are tring to do please. Maybe its a subject for wiki improvement
